src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -583,6 +583,75 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id, HR
+                                           module_metadata->assemblyName, metadata_import, nativeMethodsTypeDef);
+        if (foundType)
+        {
+            // Define the actual profiler file path as a ModuleRef
+            WSTRING native_profiler_file = GetProfilerFilePath();
+            mdModuleRef profiler_ref;
+            hr = metadata_emit->DefineModuleRef(native_profiler_file.c_str(), &profiler_ref);
+            if (SUCCEEDED(hr))
+            {
+                // Enumerate all methods inside the native methods type with the PInvokes
+                Enumerator<mdMethodDef> enumMethods = Enumerator<mdMethodDef>(
+                    [metadata_import, nativeMethodsTypeDef](HCORENUM* ptr, mdMethodDef arr[], ULONG max, ULONG* cnt) -> HRESULT
+                    {
+                        return metadata_import->EnumMethods(ptr, nativeMethodsTypeDef, arr, max, cnt);
+                    }, [metadata_import](HCORENUM ptr) -> void { metadata_import->CloseEnum(ptr); });
+
+                EnumeratorIterator<mdMethodDef> enumIterator = enumMethods.begin();
+                while (enumIterator != enumMethods.end())
+                {
+                    auto methodDef = *enumIterator;
+
+                    // Get the current PInvoke map to extract the flags and the entrypoint name
+                    DWORD pdwMappingFlags;
+                    WCHAR importName[kNameMaxSize]{};
+                    DWORD importNameLength = 0;
+                    mdModuleRef importModule;
+                    hr = metadata_import->GetPinvokeMap(methodDef, &pdwMappingFlags, importName, kNameMaxSize,
+                                                        &importNameLength, &importModule);
+                    if (SUCCEEDED(hr))
+                    {
+                        // Delete the current PInvoke map
+                        hr = metadata_emit->DeletePinvokeMap(methodDef);
+                        if (SUCCEEDED(hr))
+                        {
+                            // Define a new PInvoke map with the new ModuleRef of the actual profiler file path
+                            hr = metadata_emit->DefinePinvokeMap(methodDef, pdwMappingFlags,
+                                                                 WSTRING(importName).c_str(), profiler_ref);
+                            if (FAILED(hr))
+                            {
+                                Warn("ModuleLoadFinished: DefinePinvokeMap failed");
+                                return hr;
+                            }
+                        }
+                        else
+                        {
+                            Warn("ModuleLoadFinished: DeletePinvokeMap failed");
+                        }
+                    }
+
+                    enumIterator = ++enumIterator;
+                }
+            }
+            else
+            {
+                Warn("ModuleLoadFinished: Native Profiler DefineModuleRef failed");
```

Comment:
What are the consequences if this fails? Is there anything else we need to do?

Reply:
It shouldn't fail because is a ModuleRef definition... But if we fail here... that means the PInvokes will be broken. Currently I prefer to continue the module load and handle the exception in the PInvoke call... so all PInvoke calls should be inside a try/catch

Reply:
I'm guessing there's nothing to do because we haven't actually modified anything

Reply:
Sorry, I didn't refresh so I didn't see your response 😄 That makes sense, thanks!